File: Data Structures & Algorithms/binary-search/submission-3.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;

        while (r >= l) {
            int mid = l + (r - l) / 2;
            cout << mid << ", l: " << l << ", r: " << r << "\n";
            if (nums[mid] == target) {
                return mid;
            }

            if (target > nums[mid]) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }

        return -1;
    }
};

File: Data Structures & Algorithms/maximum-subarray/submission-3.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        nums_max = {}

        def helper(i, started):
            if i in nums_max:
                return nums_max[i]

            if i == len(nums) - 1:
                return max(0, nums[i]) if started else nums[i]
            if started:
                cur_max = max(0, nums[i] + helper(i + 1, True))
                nums_max[i] = cur_max
                return cur_max
            else:
                return max(helper(i + 1, False), nums[i] + helper(i + 1, True))
        
        return helper(0, False)
